keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1362 @@
+	// EvacuationProgressTimestamp or orphan/complete it by setting ActiveEvacuatorCompleted to true.
+	// +optional
+	ActiveEvacuatorClass *string `json:"activeEvacuatorClass,omitempty" protobuf:"bytes,1,opt,name=activeEvacuatorClass"`
+
+	// ActiveEvacuatorCompleted should be set to true when the evacuator of the
+	// ActiveEvacuatorClass has completed its full or partial evacuation.
+	// This field can also be set to true if no evacuator is available.
+	// If this field is true, no additional evacuator is available, and the evacuated pod still
+	// exists, it is evicted using the Eviction API.
+	// +optional
+	ActiveEvacuatorCompleted bool `json:"activeEvacuatorCompleted,omitempty" protobuf:"varint,2,opt,name=activeEvacuatorCompleted"`
+
+	// EvacuationProgressTimestamp is the time at which the evacuation was reported to be in progress by the evacuator.
+	// Cannot be set to the future time (after taking time skew into account).
+	// +optional
+	EvacuationProgressTimestamp *metav1.Time `json:"evacuationProgressTimestamp,omitempty" protobuf:"bytes,3,opt,name=evacuationProgressTimestamp"`
+
+	// ExpectedEvacuationFinishTime is the time at which the evacuation is expected to end for the
+	// current evacuator and its class.
+	// May be empty if no estimate can be made.
+	// +optional
+	ExpectedEvacuationFinishTime *metav1.Time `json:"expectedEvacuationFinishTime,omitempty" protobuf:"bytes,4,opt,name=expectedEvacuationFinishTime"`
+	
+	// The number of unsuccessful attempts to evict the referenced pod, e.g. due to a PDB.
+	// This field is required.
+	FailedEvictionCounter int32 `json:"failedEvictionCounter" protobuf:"varint,5,opt,name=failedEvictionCounter"`
+	
+	// Message is a human readable message indicating details about the evacuation.
+	// This may be an empty string.
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=32768
+	Message string `json:"message" protobuf:"bytes,6,opt,name=message"`
+
+	// Conditions can be used by evacuators to share additional information about the evacuation.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,7,rep,name=conditions"`
+}
+```
+
+#### Evacuation Admission
+
+Name of the Evacuation object must be unique and predictable for each pod to prevent the creation
+of multiple Evacuations for the same pod. We do not expect the evacuators to support interaction
+with multiple Evacuations for a pod.
+
+`.metadata.generateName` is not supported.
+`.metadata.name` of the pod is expected to be in the format `${POD_UID}-${POD_NAME_PREFIX}`.
+ The `POD_UID` is there for uniqueness and the `POD_NAME_PREFIX` is there for a user convenience.
+
+- The `POD_UID` (length 37 characters) must be identical to `.spec.podRef.uid` or the request will
```

Comment:
Can we guarantee that UIDs are 37 characters?

Reply:
Not sure if we can guarantee that, but we use the `NewUUID` function

- https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/staging/src/k8s.io/apiserver/pkg/registry/rest/meta.go#L38
-  https://pkg.go.dev/github.com/google/uuid#UUID.String 

which always outputs `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` which is 36 characters. 

I have fixed the off by one error.

Reply:
I don't think we make an API guarantee about the string length of this field.

Reply:
I think it should be possible to guarantee this, at least for the pods. I have added a new `Pod API` section where I mention this.